### ISSUE TITLE
Short press on the wait button somtimes waits too much

### ISSUE THREAD

**Issue by vi** [Smujb/powered-pixel-dungeon#55]
If a fast hero waits an enemy at the door, pressing "wait" should only wait for enough time for enemy to appear at the door, not also attack or walk away.

**Comment by Smujb** [comment]
Are you sure the enemy wasn't hasted or a Crab and you weren't slowed? The wait button is bound 100% of the time to "spendAndNext(1f)" with nothing complicated going on in the background. The only way this could be happening is:
- The game thought you long-pressed (could be a bug in the code or on your end)
- The Actor system is broken (very unlikely)
- The delay was not bugged, but affected by external factors eg mob/player speed factor
(note that wearing armour you don't have enough strength for contributes to the last one)

**Comment by vi** [comment]
Probably it's because of when pressing wait button I sometimes want to wait for less than 1 full move. Instead I want to wait for the next event.

This happens with the Ring of Haste more.

**Comment by Smujb** [comment]
Ah okay. Well, that'd be super difficult to program. I suppose I could try to add a way to wait half a turn, or have a customizable wait length.

**Comment by Smujb** [comment]
Keeping the issue open as a suggestion not a bug report, as it's not a bug but is potentially an interesting QoL feature

**Comment by vi** [comment]
> Well, that'd be super difficult to program.

Long-press wait typically waits until something happens, i.e. enemy is noticed.

Maybe short-press wait should be like long-press wait, but not more than 1 full move?

Another idea is to wait until current move number becomes integer (i.e. that circular thing at the top shows no fractional move status), then waiting for 1 full turn if it is already integer.

**Comment by Smujb** [comment]
That last one could certainly work. As for the first, that would still be difficult to program because of the way resting is implemented (badly)

**Comment by vi** [comment]
Can't there be resting + an "alarm clock" for 1 move ahead.

**Comment by Smujb** [comment]
Not really... ? The game just sets "resting" to true, it doesn't know when a turn has passed, needs a callback to be called from elsewhere to say an enemy has arrived.

**Comment by vi** [comment]
For example, resting automatically stops when "Starving" buff appears.

Maybe "alarm clock" can also be implemented as a hidden technical buff?

**Comment by Smujb** [comment]
Yes, but other factors affect duration of buffs and it would be unlikely to spend exactly a turn as all actors are desynced from each other. You'd spend like .999999999 or 1.00000001 of a turn.

**Comment by vi** [comment]
Shouldn't .999999999 or 1.00000001 of a turn be considered equivalent - they are both almost "1.0".

When in doubt, waiting should err on the side of waiting less rather than more.

**Comment by Smujb** [comment]
No. **No actor in the system has the same time**. The difference is entirely random and out of my control, unless I add a manual exception for it like the hero always acting first which would get messy reallly fast.

**Comment by vi** [comment]
I often notice that short-pressing the wait button makes enemies that are typically moving one tile at a time move two tiles, as if it waits for 1.5-2 full moves instead.
